agreement/player.go

```diff
@@ -260,6 +275,64 @@ func (p *player) handleCheckpointEvent(r routerHandle, e checkpointEvent) []acti
 		}}
 }
 
+func (p *player) handleWinningPayloadArrival(payload proposal, ver protocol.ConsensusVersion) {
+	// ignoring validatedAt
+	p.payloadArrivals = append(p.payloadArrivals, payload.receivedAt)
+	p.resizePayloadArrivals(ver)
+}
+
+func (p *player) resizePayloadArrivals(ver protocol.ConsensusVersion) {
+	proto := config.Consensus[ver]
+	if len(p.payloadArrivals) > proto.DynamicFilterPayloadArriavalHistory {
+		p.payloadArrivals = p.payloadArrivals[len(p.payloadArrivals)-proto.DynamicFilterPayloadArriavalHistory:]
+	}
+	for len(p.payloadArrivals) < proto.DynamicFilterPayloadArriavalHistory {
+		p.payloadArrivals = append([]time.Duration{FilterTimeout(0, ver)}, p.payloadArrivals...)
```

Comment:
We can just use a ring buffer here initialized with size `DynamicFilterPayloadArrivalHistory` and all of the values set to the default delay. Then only insert values lower than default delay. By the time we replace all of the default values we will start getting lower round times.

This avoids append/slicing and padding.

Reply:
Thanks for the comment. I changed the implementation to avoid padding, so this makes the ring construction less convenient.